core/store/src/trie/mem/resharding.rs

```diff
@@ -0,0 +1,327 @@
+    }
+}
+
+/// Recursive implementation of the algorithm of deleting keys belonging to
+/// any of the ranges given in `intervals` from the trie.
+///
+/// `root` is the root of subtree being explored.
+/// `key_nibbles` is the key corresponding to `root`.
+/// `intervals_nibbles` is the list of ranges to be deleted.
+/// `updates_tracker` track changes to the trie caused by the deletion.
+///
+/// Returns id of the node after deletion applied.
+fn delete_multi_range_recursive<'a, M: ArenaMemory>(
+    root: MemTrieNodePtr<'a, M>,
+    key_nibbles: Vec<u8>,
+    intervals_nibbles: &[Range<Vec<u8>>],
+    updates_tracker: &mut UpdatesTracker,
+) -> Option<OldOrUpdatedNodeId> {
+    let decision = delete_decision(&key_nibbles, intervals_nibbles);
+    match decision {
+        DeleteDecision::Skip => return Some(OldOrUpdatedNodeId::Old(root.id())),
+        DeleteDecision::DeleteAll => {
+            return None;
+        }
+        DeleteDecision::Descend => {}
+    }
+
+    let node_view = root.view();
+
+    let mut resolve_branch = |children: &ChildrenView<'a, M>, mut value: Option<&ValueView>| {
+        let mut new_children = [None; 16];
+        let mut changed = false;
+
+        if intervals_nibbles.iter().any(|interval| interval.contains(&key_nibbles)) {
+            value = None;
+            changed = true;
+        }
+
+        for i in 0..16 {
+            if let Some(child) = children.get(i) {
+                let child_key_nibbles = [key_nibbles.clone(), vec![i as u8]].concat();
+                let new_child = delete_multi_range_recursive(
+                    child,
+                    child_key_nibbles,
+                    intervals_nibbles,
+                    updates_tracker,
+                );
+                match new_child {
+                    Some(OldOrUpdatedNodeId::Old(id)) => {
+                        new_children[i] = Some(OldOrUpdatedNodeId::Old(id));
+                    }
+                    Some(OldOrUpdatedNodeId::Updated(id)) => {
+                        changed = true;
+                        new_children[i] = Some(OldOrUpdatedNodeId::Updated(id));
+                    }
+                    None => {
+                        changed = true;
+                        new_children[i] = None;
+                    }
+                }
+            }
+        }
+
+        if changed {
+            let new_node = UpdatedMemTrieNode::Branch {
+                children: Box::new(new_children),
+                value: value.map(|v| v.to_flat_value()),
+            };
+            // TODO(#12074): squash the branch if needed.
+            // TODO(#12074): return None if needed.
+            Some(OldOrUpdatedNodeId::Updated(add_node(updates_tracker, new_node)))
+        } else {
+            Some(OldOrUpdatedNodeId::Old(root.id()))
+        }
+    };
+
+    match node_view {
+        MemTrieNodeView::Leaf { extension, .. } => {
+            let extension = NibbleSlice::from_encoded(extension).0;
```

Comment:
There are a lot of places using `NibbleSlice::from_encoded(...).0` and I wouldn't touch them for now, seems better to fix them all at once.
Also not sure if a getter method is applicable because from_encoded returns (Self, bool).